experimental/juvix/contract.jvx

```diff
@@ -0,0 +1,198 @@
+Map = Data.Map
+
+-- common module interfaces for data structures in Common?
+MonoidMeasure : Common.Measure
+MonoidMeasure =
+  type T = Nat
+
+  -- (<>) is normal semi-group  type class function
+  (<>)    = (<>)
+  measure = id
+
+
+-- need OrdMap, as modules that one goes to can't be modules?
+Accounts = Map.OrdMap Data.String MonoidMeasure
+
+-- we send in the string module, as it has a total order
+-- function named compare, that is also part of a general type class
+Token =
+  type Address  = s : String.T {String.length s == 36}
+
+  type Storage = {
+    total-supply : Nat.T;
+    accounts     : Accounts.T { Accounts.measure-value == total-supply }
+  }
+
+  empty-storage : Storage
+  empty-storage = {
+    total-supply = 0;
+    accounts     = Accounts.empty;
+  }
+
+  type T = {
```

Comment:
Is this a record without a constructor? Seems like we might want a named constructor.

Reply:
I would probably argue, that we shouldn't have one.

This is because if we add a constructor then, we cut off the ability to abstract. Imagine an example like this


```ocaml

type Foo = {
  i : int;
  b : string;
}

type Data1 = 
  | Foo Foo
  | B Int  

type Blah = 
  | Blah Foo
  | C String

f { i ; b } = to_string i ++ b

data1 = function
  | Foo foo = f foo
  | B i = to_string i

blah = function
  | Blah foo = f foo
  | c string = string
```

Reply:
The questions are:
- How do we deal with two records with the same fields but distinct types?
- Do we want row polymorphism?
- Can we get row polymorphism without type theory alterations?

Reply:
Here is an example from OCaml
```ocaml

utop # type foo = {a : int};;
type foo = { a : int; }

utop # type bah = {a : int};;
type bah = { a : int; }

utop # {a = 3};;
- : bah = {a = 3}

utop # let foo : foo = {a = 3};;
val foo : foo = {a = 3}

utop # let bah : bah = {a = 3};;
val bah : bah = {a = 3}
utop # foo == bah;;
Error: This expression has type bah but an expression was expected of type foo

utop # bah = { a = 3 };;
- : bool = true


utop # foo = { a = 3 };;
- : bool = true

```

It seems to allow the creation of records with the same fields.

It infers the last one made (a sane choice!)

It then disallows them being the same.

However thanks to nice type inference, if we make one on the spot it is the correct type.

Reply:
Right, that looks pretty reasonable. Ocaml doesn't have row polymorphism, though, should we consider that as well? What would it require?

Reply:
I think I am in agreement on the strategy here but this should be in a design doc (this comment thread, basically).

Reply:
> Right, that looks pretty reasonable. Ocaml doesn't have row polymorphism, though, should we consider that as well? What would it require?

Acutally OCaml does!

Here is a passage from Real World Ocaml.

"Objects have some advantages over records: they don’t require type definitions, and their support for row polymorphism makes them more flexible. However, the heavy syntax and additional runtime cost means that objects are rarely used in place of records"

Reply:
Interesting. What is the additional runtime cost? That sounds problematic, we definitely want to avoid additional runtime costs.

Reply:
The additional runtime costs aren't due to row polymorphism, it's the runtime cost of the object system (dynamic calls).

We use simulated row polymorphism all the time in Juvix, via capabilities

Reply:
Right right, that's fine then.